eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,595 @@
+	if opcode == vm.REVERT || opcode == vm.RETURN {
+		if len(t.lastThreeOpCodes) > 0 {
+			t.lastThreeOpCodes = t.lastThreeOpCodes[len(t.lastThreeOpCodes)-1:]
+		}
+	}
+}
+
+func (t *erc7562Tracer) handleGasObserved(lastOp vm.OpCode, opcode vm.OpCode, currentCallFrame *callFrameWithOpcodes) {
+	// [OP-012]
+	isCall := opcode == vm.CALL || opcode == vm.STATICCALL || opcode == vm.DELEGATECALL || opcode == vm.CALLCODE
+	pendingGasObserved := lastOp == vm.GAS && !isCall
+	if pendingGasObserved {
+		//currentCallFrame.GasObserved = true
+		incrementCount(currentCallFrame.UsedOpcodes, vm.GAS)
+	}
+}
+
+func (t *erc7562Tracer) storeUsedOpcode(opcode vm.OpCode, currentCallFrame *callFrameWithOpcodes) {
+	// ignore "unimportant" opcodes
+	if opcode != vm.GAS && !t.isIgnoredOpcode(opcode) {
+		incrementCount(currentCallFrame.UsedOpcodes, opcode)
+	}
+}
+
+func (t *erc7562Tracer) handleStorageAccess(opcode vm.OpCode, scope tracing.OpContext, currentCallFrame *callFrameWithOpcodes) {
+	if opcode == vm.SLOAD || opcode == vm.SSTORE || opcode == vm.TLOAD || opcode == vm.TSTORE {
+		slot := common.BytesToHash(peepStack(scope.StackData(), 0).Bytes())
+		slotHex := slot.Hex()
+		addr := scope.Address()
+
+		if opcode == vm.SLOAD {
+			// read slot values before this UserOp was created
+			// (so saving it if it was written before the first read)
+			_, rOk := currentCallFrame.AccessedSlots.Reads[slotHex]
+			_, wOk := currentCallFrame.AccessedSlots.Writes[slotHex]
+			if !rOk && !wOk {
+				currentCallFrame.AccessedSlots.Reads[slotHex] = append(currentCallFrame.AccessedSlots.Reads[slotHex], t.env.StateDB.GetState(addr, slot).Hex())
+			}
+		} else if opcode == vm.SSTORE {
+			incrementCount(currentCallFrame.AccessedSlots.Writes, slotHex)
+		} else if opcode == vm.TLOAD {
+			incrementCount(currentCallFrame.AccessedSlots.TransientReads, slotHex)
+		} else if opcode == vm.TSTORE {
+			incrementCount(currentCallFrame.AccessedSlots.TransientWrites, slotHex)
+		}
+	}
+}
+
+func (t *erc7562Tracer) storeKeccak(opcode vm.OpCode, scope tracing.OpContext) {
+	if opcode == vm.KECCAK256 {
+		dataOffset := peepStack(scope.StackData(), 0).Uint64()
+		dataLength := peepStack(scope.StackData(), 1).Uint64()
+		memory := scope.MemoryData()
+		keccak := make([]byte, dataLength)
+		copy(keccak, memory[dataOffset:dataOffset+dataLength])
+		t.Keccak = append(t.Keccak, keccak)
+	}
+}
+
+func (t *erc7562Tracer) detectOutOfGas(gas uint64, cost uint64, opcode vm.OpCode, currentCallFrame *callFrameWithOpcodes) {
+	if gas < cost || (opcode == vm.SSTORE && gas < 2300) {
+		currentCallFrame.OutOfGas = true
+	}
+}
+
+func (t *erc7562Tracer) handleExtOpcodes(lastOpInfo *opcodeWithPartialStack, opcode vm.OpCode, currentCallFrame *callFrameWithOpcodes) {
+	if isEXT(lastOpInfo.Opcode) {
+		addr := common.HexToAddress(lastOpInfo.StackTop3[0].Hex())
+
+		// TODO: THIS CODE SEEMS TO BE EQUIVALENT TO THE STRING BASED BUT IT IS LOGICALLY WRONG AND WILL NEVER WORK.
```

Comment:
Did you check that this code is actually correct? Let's write a comment explaining it, maybe, as I find it confusing.

Reply:
I did, it's Ok. It was buggy since we passed stackTop3 of the current opcode instead of the lastOp, and worse, we kept a pointer to the stack in each `opcodeWithPartialStack`. But now It works fine.